File: backend/src/test/java/edonymyeon/backend/member/domain/EmailTest.java
package edonymyeon.backend.member.domain;

import static edonymyeon.backend.global.exception.ExceptionInformation.MEMBER_EMAIL_INVALID;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import edonymyeon.backend.global.exception.EdonymyeonException;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayNameGeneration;
import org.junit.jupiter.api.DisplayNameGenerator;
import org.junit.jupiter.api.Test;

@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
@SuppressWarnings("NonAsciiCharacters")
class EmailTest {

    @Test
    void 이메일을_String으로_생성한다() {
        //given
        final String email = "email@test.com";

        //when
        //then
        Assertions.assertDoesNotThrow(() -> Email.from(email));
    }

    @Test
    void 이메일을_SocialType으로_생성한다() {
        //given
        final SocialInfo.SocialType socialType = SocialInfo.SocialType.KAKAO;

        //when
        //then
        Assertions.assertDoesNotThrow(() -> Email.from(socialType));
    }

    @Test
    void 이메일이_30글자를_초과하면_예외가_발생한다() {
        //given
        StringBuilder stringBuilder = new StringBuilder();
        for (int i = 0; i < 31; i++) {
            stringBuilder.append(i);
        }
        final String email = stringBuilder.toString();

        //when
        //then
        assertThatThrownBy(() -> Email.from(email))
                .isInstanceOf(EdonymyeonException.class)
                .hasMessage(MEMBER_EMAIL_INVALID.getMessage());
    }
}

File: android/domain/src/test/kotlin/com/domain/edonymyeon/model/EmailTest.kt
package com.domain.edonymyeon.model

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertDoesNotThrow
import org.junit.jupiter.api.assertThrows

internal class EmailTest {
    @Test
    fun `이메일 형식으로 생성한다`() {
        assertDoesNotThrow {
            Email.create("abc@naver.com")
        }
    }

    @Test
    fun `이메일 형식이 아니면 예외가 발생한다`() {
        assertThrows<IllegalArgumentException> {
            Email.create("abc@")
        }
    }

    @Test
    fun `이메일에 공백을 넣으면 제거된다`() {
        val email = Email.create("abc@naver.com ")
        assertThat(email.value).isEqualTo("abc@naver.com")
    }

    @Test
    fun `이메일 검증에서 이메일 형식이면 true가 반환된다`() {
        val result = Email.validate("abc@naver.com")
        assertThat(result).isEqualTo(true)
    }

    @Test
    fun `이메일 검증에서 이메일 형식이 아니면 false가 반환된다`() {
        val result = Email.validate("abc")
        assertThat(result).isEqualTo(false)
    }
}
